server/src/main/java/org/elasticsearch/index/engine/RecoverySourcePruneMergePolicy.java

```diff
@@ -0,0 +1,286 @@
+    RecoverySourcePruneMergePolicy(String recoverySourceField, Supplier<Query> retainSourceQuerySupplier, MergePolicy in) {
+        super(in, toWrap -> new OneMerge(toWrap.segments) {
+            @Override
+            public CodecReader wrapForMerge(CodecReader reader) throws IOException {
+                CodecReader wrapped = toWrap.wrapForMerge(reader);
+                return wrapReader(recoverySourceField, wrapped, retainSourceQuerySupplier);
+            }
+        });
+
+    }
+
+    // pkg private for testing
+    static CodecReader wrapReader(String recoverySourceField, CodecReader reader, Supplier<Query> retainSourceQuerySupplier)
+        throws IOException {
+        NumericDocValues recovery_source = reader.getNumericDocValues(recoverySourceField);
+        if (recovery_source == null || recovery_source.nextDoc() == DocIdSetIterator.NO_MORE_DOCS) {
+            return reader; // early terminate - nothing to do here since non of the docs has a recovery source anymore.
+        }
+        BooleanQuery.Builder builder = new BooleanQuery.Builder();
+        builder.add(new DocValuesFieldExistsQuery(recoverySourceField), BooleanClause.Occur.FILTER);
+        builder.add(retainSourceQuerySupplier.get(), BooleanClause.Occur.FILTER);
+        IndexSearcher s = new IndexSearcher(reader);
+        s.setQueryCache(null);
+        Weight weight = s.createWeight(builder.build(), false, 1.0f);
+        Scorer scorer = weight.scorer(reader.getContext());
+        if (scorer != null) {
+            return new SourcePruningFilterCodecReader(recoverySourceField, reader, BitSet.of(scorer.iterator(), reader.maxDoc()));
+        } else {
+            return new SourcePruningFilterCodecReader(recoverySourceField, reader, new BitSet.MatchNoBits(reader.maxDoc()));
+        }
+    }
+
+    private static class SourcePruningFilterCodecReader extends FilterCodecReader {
+        private final Bits recoverySourceToKeep;
+        private final String recoverySourceField;
+
+        SourcePruningFilterCodecReader(String recoverySourceField, CodecReader reader, Bits recoverySourceToKeep) {
+            super(reader);
+            this.recoverySourceField = recoverySourceField;
+            this.recoverySourceToKeep = recoverySourceToKeep;
+        }
+
+        @Override
+        public DocValuesProducer getDocValuesReader() {
+            DocValuesProducer docValuesReader = super.getDocValuesReader();
+            return new FilterDocValuesProducer(docValuesReader) {
+                @Override
+                public NumericDocValues getNumeric(FieldInfo field) throws IOException {
+                    NumericDocValues numeric = super.getNumeric(field);
+                    if (recoverySourceField.equals(field.name)) {
+                        assert numeric != null : recoverySourceField + " must have numeric DV but was null";
+                        return new FilterNumericDocValues(numeric) {
```

Comment:
If `recoverySourceToKeep` was a bitset, we could do a leap frog, which would be faster if `recoverySourceToKeep` is sparse.

```java
final ConjunctionDISI intersection = ConjunctionDISI.intersectIterators(Arrays.asList(numeric, new BitSetIterator(recoverySourceToKeep)));
return new FilterNumericDocValues(numeric) {
  @Override
  public int nextDoc() throws IOException {
    return intersection.nextDoc();
  }
};
```